ProjFS.Mac/PrjFSKext/Message_Kernel.cpp

```diff
@@ -33,7 +35,57 @@ void Message_Init(
     {
         header->pathSizeBytes = 0;
     }
-    
+
+    if (nullptr != fromPath)
+    {
+        header->fromPathSizeBytes = strlen(fromPath) + 1;
+    }
+    else
+    {
+        header->fromPathSizeBytes = 0;
+    }
+
     spec->messageHeader = header;
     spec->path = path;
+    spec->fromPath = fromPath;
+}
+
+uint32_t Message_EncodedSize(const Message& message)
+{
+    return sizeof(*message.messageHeader) + message.messageHeader->pathSizeBytes + message.messageHeader->fromPathSizeBytes;
+}
+
+uint32_t Message_Encode(void* buffer, const uint32_t bufferSize, const Message& message)
+{
+    uint8_t* bufferPosition = static_cast<uint8_t*>(buffer);
+    uint32_t bufferBytesRemain = bufferSize;
+    
+    assert(bufferSize >= sizeof(*message.messageHeader));
+    memcpy(bufferPosition, message.messageHeader, sizeof(*message.messageHeader));
+    bufferPosition +=    sizeof(*message.messageHeader);
+    bufferBytesRemain -= sizeof(*message.messageHeader);
+    
+    {
```

Comment:
For my own info, why is this done in a set of {'s

Reply:
Re: all the questions/comments about this function - the code of this function is still in flux, this is a prototyping iteration on the original loop-based approach. I don't like the current approach with the repeated code, so I'm going to find a better solution to the overall structure of the code before picking apart the style.